api/provider/notion.go

```diff
@@ -0,0 +1,111 @@
+package provider
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/netlify/gotrue/conf"
+	"golang.org/x/oauth2"
+)
+
+const defaultNotionApiBase 	= "api.notion.com"
+
+type notionProvider struct {
+	*oauth2.Config
+	APIPath string
+}
+
+type notionUser struct {
+	Bot struct {
+		Owner struct {
+			User struct {
+				ID        string `json:"id"`
+				Name      string `json:"name"`
+				AvatarURL string `json:"avatar_url"`
+				Person struct {
+					Email     string `json:"email"`
+				} `json:"person"`
+			} `json:"user"`
+		} `json:"owner"`
+	} `json:"bot"`
+}
+
+// NewNotionProvider creates a Notion account provider.
+func NewNotionProvider(ext conf.OAuthProviderConfiguration) (OAuthProvider, error) {
+	if err := ext.Validate(); err != nil {
+		return nil, err
+	}
+
+	authHost := chooseHost(ext.URL, defaultNotionApiBase)
+
+	return &notionProvider{
+		Config: &oauth2.Config{
+			ClientID:     ext.ClientID,
+			ClientSecret: ext.Secret,
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  authHost + "/v1/oauth/authorize",
+				TokenURL: authHost + "/v1/oauth/token",
+			},
+			RedirectURL: ext.RedirectURI,
+		},
+		APIPath: authHost,
+	}, nil
+}
+
+func (g notionProvider) GetOAuthToken(code string) (*oauth2.Token, error) {
+	return g.Exchange(oauth2.NoContext, code)
+}
+
+func (g notionProvider) GetUserData(ctx context.Context, tok *oauth2.Token) (*UserProvidedData, error) {
+	var u notionUser
+
+	// Perform http request, because we need to set the Notion-Version header
+	req, err := http.NewRequest("GET", g.APIPath + "/v1/users/me", nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// set headers
+	req.Header.Set("Notion-Version", "2021-08-16")
```

Comment:
Hey! 

Not a maintainer for this repo but wondering if the date `2021-08-16` will need to be modified when Notion releases a new ver.

Should we put a comment or declare it as a constant somewhere?

Reply:
Hi @J0 

Yup the version need to be modified when Notion release new version of their API. 

Ideally the users can select which version of API they want to use by passing  `{ "version": "2021-08-16" }`, but seeing that supabase-js only has [limited options](https://supabase.com/docs/reference/javascript/auth-signin#parameters), so I set it manually. 

Adding a comment is a great suggestion! Thank you! 🤩

Reply:
Hey @zernonia,

Also agreeing with @J0, I think a constant might be best - just like you have a constant named `defaultNotionApiBase`  you could have one like `notionApiVersion`, then you might not even need a comment because the code would be self-documenting

Reply:
Thank you for the suggestions. @J0  @i-pip 

I've added Notion Version to the constant